### PR TITLE
Refactor chain metadata

### PR DESCRIPTION
**Pull request by jmrossy**
### Description

- Merge the chain metadata from the explorer, the existing sdk metadata, and the sdk's chain connection configs, all into one
- De-dupe chain finality configs
- Alphabetize chain names and configs
- Fix circular dep in types
- Add convenience utils for retrieving chain metadata

### Related issues

Pre-req for #1322 

### Backward compatibility

**No**: The new metadata schema is different. To ease migration, the old one has been included (and marked deprecated) under the name `partialChainMetadata`

### CODE REVIEW

**Review by nambrot — APPROVED** [review]
Lgtm, but I would have @tkporter take a look as well in regards to agent configuration

**Comment by asaj** [comment]
Yeah I'm guessing `reorgPeriod` in infra/config/testnet3/agent.ts should be included?

File: typescript/sdk/src/consts/chainMetadata.ts [hunk 1]
```diff
@@ -1,187 +1,606 @@
...
+  /** Collection of block explorers */
+  blockExplorers: Array<{
+    name: string;
+    url: string;
+    family: ExplorerFamily;
+    apiUrl?: string;
+  }>;
+  blocks: {
+    // Number of blocks to be considered final
+    finalityConfirmations: number;
```
Comment (asaj) [inline typescript/sdk/src/consts/chainMetadata.ts:45]: I wonder if this should just be "confirmations: Number of blocks to wait before considering a transaction confirmed"

If you add reorgPeriod (or alternatively "finality"), consider "The number of blocks before a transaction has a near zero chance of reverting"

Comment (jmrossy) [reply typescript/sdk/src/consts/chainMetadata.ts:45]: Yep that sgtm

File: typescript/sdk/src/consts/chainConnectionConfigs.ts [hunk 2]
```diff
@@ -3,206 +3,40 @@ import { ethers } from 'ethers';
...
+      'http://localhost:8545',
+      31337,
+    ),
+    confirmations: 1,
+  };
+}
+
+export const chainConnectionConfigs: ChainMap<ChainName, IChainConnection> =
+  objMap(chainMetadata, (chainName, metadata) => {
+    if (chainName.includes('test')) return testChainConnection();
```
Comment (tkporter) [inline typescript/sdk/src/consts/chainConnectionConfigs.ts:23]: I think this condition may hit for chains like `bsctestnet`. Maybe a more strict condition is needed to ensure only test1/test2/test3 match

File: typescript/infra/config/environments/testnet3/agent.ts [hunk 3]
```diff
@@ -44,22 +46,22 @@ export const hyperlane: AgentConfig<TestnetChains> = {
     },
     chainOverrides: {
       alfajores: {
-        reorgPeriod: 0,
+        reorgPeriod: chainMetadata.alfajores.blocks.reorgPeriod,
```
Comment (tkporter) [inline typescript/infra/config/environments/testnet3/agent.ts:49]: love this, much nicer

**Review by tkporter — APPROVED** [review]
lgtm!

File: typescript/sdk/src/consts/chainConnectionConfigs.ts (hunk 2 above)
Comment (jmrossy) [reply typescript/sdk/src/consts/chainConnectionConfigs.ts:23]: Oh great catch, thank you!

**Pull request merged**